### PR TITLE
Fixes #1542: avoid blocking fetch in C* store

### LINKED ISSUE #1542: Cassandra3 span store randomly throws exceptions when retrieving spans.
Zipkin Version: 1.20.0
Cassandra Version: 3.9

The Cassandra3 CassandraSpanStore.getSpansByTraceIds randomly throws the following exception.

The underlying issue seems to be that the driver sees that the result set current page is exhausted and prepares to fetch more results while in an asynchronous request.  When this happens the driver makes a check to see if the synchronous request is being made while in an event loop.  Then end result is the following stack trace:

```
java.lang.IllegalStateException: Detected a synchronous call on an I/O thread, this can cause deadlocks or unpredictable behavior. This generally happens when a Future callback calls a synchronous Session method (execute() or prepare()), or iterates a result set past the fetch size (causing an internal synchronous fetch of the next page of results). Avoid this in your callbacks, or schedule them on a different executor.
         at com.datastax.driver.core.AbstractSession.checkNotInEventLoop(AbstractSession.java:203) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.ArrayBackedResultSet$MultiPage.prepareNextRow(ArrayBackedResultSet.java:310) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.ArrayBackedResultSet$MultiPage.isExhausted(ArrayBackedResultSet.java:269) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.ArrayBackedResultSet$1.hasNext(ArrayBackedResultSet.java:143) ~[cassandra-driver-core-3.1.3.jar!/:na]
        at zipkin.storage.cassandra3.CassandraSpanStore$4.apply(CassandraSpanStore.java:354) ~[zipkin-storage-cassandra3-1.20.0.jar!/:na]
         at zipkin.storage.cassandra3.CassandraSpanStore$4.apply(CassandraSpanStore.java:351) ~[zipkin-storage-cassandra3-1.20.0.jar!/:na]
         at com.google.common.util.concurrent.Futures$2.apply(Futures.java:760) ~[guava-18.0.jar!/:na]
         at com.google.common.util.concurrent.Futures$ChainingListenableFuture.run(Futures.java:906) ~[guava-18.0.jar!/:na]
         at zipkin.autoconfigure.storage.cassandra3.brave.TracedSession$BraveResultSetFuture.lambda$addListener$0(TracedSession.java:201) ~[zipkin-autoconfigure-storage-cassandra3-1.20.0.jar!/:na]
         at zipkin.autoconfigure.storage.cassandra3.brave.TracedSession$BraveResultSetFuture$$Lambda$1.run(Unknown Source) ~[zipkin-autoconfigure-storage-cassandra3-1.20.0.jar!/:na]
         at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457) ~[guava-18.0.jar!/:na]
         at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156) ~[guava-18.0.jar!/:na]
         at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145) ~[guava-18.0.jar!/:na]
         at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185) ~[guava-18.0.jar!/:na]
         at com.datastax.driver.core.DefaultResultSetFuture.onSet(DefaultResultSetFuture.java:174) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.RequestHandler.setFinalResult(RequestHandler.java:177) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.RequestHandler.access$2500(RequestHandler.java:46) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.RequestHandler$SpeculativeExecution.setFinalResult(RequestHandler.java:797) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:500) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1070) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:993) ~[cassandra-driver-core-3.1.3.jar!/:na]
         at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:328) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:321) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:266) ~[netty-handler-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:328) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:321) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102) ~[netty-codec-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:328) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:321) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293) ~[netty-codec-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267) ~[netty-codec-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:328) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:321) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1280) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:328) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:890) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:564) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:505) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:419) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:391) ~[netty-transport-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112) ~[netty-common-4.0.37.Final.jar!/:4.0.37.Final]
         at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145) ~[netty-common-4.0.37.Final.jar!/:4.0.37.Final]
         at java.lang.Thread.run(Thread.java:745) [na:1.8.0_112]
```

### PR DESCRIPTION
**Pull request by llinder**
The Cassandra driver performs a blocking call when fetching more results.  Async executions prohibit blocking calls while in an async operation.  When the driver detects this it throws an exception.  The recommended approach to handling page fetches while in an async request is to chain a new request when the page set is exhausted.

/cc @openzipkin/cassandra

### CODE REVIEW

File: zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java [hunk 1]
```diff
@@ -468,21 +456,101 @@
...
       return immediateFailedFuture(ex);
     }
   }
+
+  private static <K, T> AsyncFunction<ResultSet, Map<K, T>> readResultsAsMap(final Map<K, T> results,
+      Function<Row, Map.Entry<K, T>> rowMapper) {
+    return new AsyncFunction<ResultSet, Map<K, T>>() {
+      @Override public ListenableFuture<Map<K, T>> apply(ResultSet rs) throws Exception {
+        // How far we can go without triggering the blocking fetch:
+        int remainingInPage = rs.getAvailableWithoutFetching();
```
Comment (codefromthecrypt) [inline zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java:470]: how many is this usually. I wonder if we can change our tests to make sure this is invoked?

Comment (llinder) [reply zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java:470]: The default fetch size is 5000.  I think there are 2 ways to try and test, either set fetch size lower or add more data.  Setting the fetch size lower is a bit harder but we could try doing that when the driver is instantiated.

**Comment by codefromthecrypt** [comment]
Eventhough 5k sounds like a lot. This is deduping client side right? In
your site what is the amount of names returned vs unique?

**Comment by llinder** [comment]
The case where we hit this the most was on `getSpansByTraceIds`.  If we put a high number, sometimes even 1000, in the UI trace limit field it can trigger the error.  For other read paths such as `getSpanNames` we didn't hit this error.  Even though we didn't encounter this in other read paths, the pattern seemed easy enough to generalize that it could be applied to all areas and not just `getSpansByTraceIds`.

**Comment by codefromthecrypt** [comment]
>
> The case where we hit this the most was on getSpansByTraceIds. If we put
> a high number, sometimes even 1000, in the UI trace limit field it can
> trigger the error. For other read paths such as getSpanNames we didn't
> hit this error. Even though we didn't encounter this in other read paths,
> the pattern seemed easy enough to generalize that it could be applied to
> all areas and not just getSpansByTraceIds.
>
ah thx for the insight. so basically we should be able to find some command
that creates this error, even if it is a sneaky test setup.

**Comment by llinder** [comment]
Testing this locally I was able to trigger it by increasing the trace count here: https://github.com/openzipkin/zipkin/blob/master/zipkin-storage/cassandra3/src/test/java/zipkin/storage/cassandra3/CassandraSpanStoreTest.java#L71

I was hesitant to break 5000 traces just for the test.  Ideally we could set the default fetch size on the driver to a lower value and it will trigger this error as well.  Setting the fetch size for the tests will take some work though and possibly some changes to the Cassandra3Storage builder.

**Comment by codefromthecrypt** [comment]
Mind trying maybe with TestObjects.LOTS_OF_SPANS sliced at 5001
(instead of the old code that replicates LOTS_OF_SPANS)? If it doesn't
dramatically increase the test duration, should be fine.

**Comment by llinder** [comment]
Sure i will try increasing that tomorrow.

**Comment by codefromthecrypt** [comment]
merging this as cassandra3 is experimental and leaving this unmerged keeps the code broken. Will try and merge in the other things so we can follow-up with tests.

File: zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java (hunk 1 above)
Comment (michaelsembwever) [reply zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java:470]: I'm a bit late to the party on this one.

Would it not be simpler to call `rs.fetchMoreResults()` ?

`fetchMoreResults()` is an async call, and improves the chances that the next page is ready to go by the time this page is exhausted.

Taking this approach is simpler as it removes the concurrency/futures from this code, allowing it to be just the single iteration loop

Comment (codefromthecrypt) [reply zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java:470]: opened https://github.com/openzipkin/zipkin/issues/1647 to track. thanks!

Comment (llinder) [reply zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java:470]: @michaelsembwever thanks for taking a look at this.  I'm not sure I completely follow your suggestion though.  Is there a reference example that you can point me at?

The approach I've taken here was borrowed from Datastax documentation on async paging https://docs.datastax.com/en/developer/java-driver/3.1/manual/async/#async-paging

**Pull request merged**